src/app/business/delfin/storages/storages.component.ts

```diff
@@ -0,0 +1,601 @@
+        _.each(groupedData, function(value, key){
+            let parentTreeNode = {
+                label: key,
+                collapsedIcon: 'fa-folder',
+                expandedIcon: 'fa-folder-open',
+                type: 'array'
+            }
+            // Check if if the grouped devices exist and prepare the second level of nodes as children.
+            if(isArray(value) && value.length){
+                parentTreeNode['children'] = [];
+                _.each(value, function(storageDevice){
+                    let parentTreeNodeChild = {
+                        label: storageDevice['name'],
+                        collapsedIcon: 'fa-hdd-o',
+                        expandedIcon: 'fa-hdd-o',
+                        children: [],
+                        type: 'device',
+                        details: storageDevice
+                    }
+                    // Populate Volumes children
+                    let parentVolNode = {};
+                    if(storageDevice['volumes'] && storageDevice['volumes'].length){
+                        parentVolNode = {
+                            label : "Volumes",
+                            collapsedIcon: 'fa-database',
+                            expandedIcon: 'fa-database',
+                            type: 'volParent',
+                            children: []
+                        }
+                        parentTreeNodeChild['children'].push(parentVolNode);
+                    }
+                    
+                    // Push each storage device as a child of the vendor node
+                    parentTreeNode['children'].push(parentTreeNodeChild);
+                });
+            }
+            // Push each Vendor as a node in the Tree
+            treeData.push(parentTreeNode);
+        })
+        this.arrayTreeData = treeData;
+        
+    }
+
+    lazyLoadTreeChildren(event){
+
+    }
+
+    //Show the overview panel when hovering on device in the tree.
+    showOverview(event, storage, overlaypanel: OverlayPanel){
+        this.storageOverview = storage;
+        overlaypanel.toggle(event);
+    }
+
+    getModelsByVendor(vendor){
+        let self =this;
+        _.each(this.allStorageModels, function(value, key){
+            if(key==vendor){
+                self.modelOptions = value;
+            }
+        })
+    }
+
+    returnSelectedStorage(selectedStorage){
+        this.selectStorage = selectedStorage;
+    }
+
+    batchDeleteStorages(storages){
+        if(storages){
+            let  msg, arr = [], selectedNames=[];
+            if(_.isArray(storages)){
+                storages.forEach((item,index)=> {
+                    arr.push(item.id);
+                    selectedNames.push(item['name']);
+                })
+                msg = "<h3>Are you sure you want to delete the selected " + storages.length + " Storages?</h3><h4>[ "+ selectedNames.join(',') +" Storage(s) ]</h4>";
+            }else{
+                arr.push(storages.id)
```

Comment:
Why the batch delete storage will fall in this situation? It is internal call and user directly calls this, so is this part of code required?

Reply:
For deleting a storage device there are two ways. 
One is by clicking on `More` in the operations  menu and clicking on the delete button for a particular device.
The other method is by using the checkbox to select multiple devices.
Here I reuse the `batchDeleteStorages()` method in both cases. 
In case the object received is an array then pick the `id`, push into the `arr` array and display the names to the user to confirm delete.
In case it is an object and not an array then push the `id` into the `arr` array and show the appropriate message to the user.
In both cases an array of objects `arr` is sent for deletion.
If multiple selected then its an array of multiple objects else its an array of one object

Reply:
Ok